pkg/operator/etcdmemberlistercontroller/etcdmemberlistercontroller.go

```diff
@@ -0,0 +1,195 @@
+package etcdmemberlistercontroller
+
+import (
+	"context"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+
+	errorsutil "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-etcd-operator/pkg/dnshelpers"
+	"github.com/openshift/cluster-etcd-operator/pkg/etcdcli"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+var (
+	errNilMembersCache = errors.New("member cache is nil")
+	errKeyExist        = errors.New("key already exist")
+	errKeyNotExist     = errors.New("key does not exist")
+)
+
+type membersCache struct {
+	mux   sync.RWMutex
+	cache map[string]etcdserverpb.Member
+}
+
+func newMembersCache() *membersCache {
+	return &membersCache{
+		mux:   sync.RWMutex{},
+		cache: make(map[string]etcdserverpb.Member),
+	}
+}
+
+func (mc *membersCache) add(key string, value *etcdserverpb.Member) (bool, error) {
+	if mc == nil || mc.cache == nil {
+		return false, errNilMembersCache
+	}
+	defer mc.mux.Unlock()
+
+	mc.mux.Lock()
+	_, ok := mc.cache[key]
+	if ok {
+		return false, errKeyExist
+	}
+	// safe to add
+	mc.cache[key] = *value
+	return true, nil
+}
+
+func (mc *membersCache) update(key string, value *etcdserverpb.Member) (bool, error) {
+	if mc == nil || mc.cache == nil {
+		return false, errNilMembersCache
+	}
+	defer mc.mux.Unlock()
+
+	mc.mux.Lock()
+	_, ok := mc.cache[key]
+	if !ok {
+		return false, errKeyNotExist
+	}
+	// safe to update
+	mc.cache[key] = *value
+	return true, nil
+}
```

Comment:
Do we really need add/update/delete/get methods on the member cache if we're just outright replacing the cache when the membership changes?
If not, then we can remove these.

Reply:
yes u r right, initially i was using them to update the cache with new entries and remove stale entries .. but we dont need them now, removing ...